SI-8306: handle SWITCH nodes with just default case

gkossakowski opened the pull request:
Handle properly SWITCH nodes that contain just a default case in optimizer
(ConstantOptimization). SWITCH with just default case is expressed as a
node with empty tags and exactly one label. We can handle such nodes
easily by introducing a shortcut in logic that computes reachableLabels.

Add a test case which triggers patmat to generate SWITCH node with just
a default case.

Fixes SI-8306.

gkossakowski commented:
Review by @retronym.

retronym commented on test/files/pos/SI-8306.scala:
```diff
@@ -0,0 +1,8 @@
+class Si8306 {
```
File should be called `t8306`. The indentation is off, too.

retronym commented on test/files/pos/SI-8306.scala:
```diff
@@ -0,0 +1,8 @@
+class Si8306 {
+	def foo: Int = 123
+	lazy val extension: Int =
```
I wonder why lazy mattered.

gkossakowski committed:
SI-8306: handle SWITCH nodes with just default case

Handle properly SWITCH nodes that contain just a default case in optimizer
(ConstantOptimization). SWITCH with just default case is expressed as a
node with empty tags and exactly one label. We can handle such nodes
easily by introducing a shortcut in logic that computes reachableLabels.

Add a test case which triggers patmat to generate SWITCH node with just
a default case.

Fixes SI-8306.

gkossakowski commented:
Fixed file name and indentation.

I'll try to see why lazy val matters here.

retronym commented:
LGTM

gkossakowski commented:
@retronym: you asked why `lazy` mattered in this case. It's pure coincidence that `lazy val` happened to trigger the problem. You can just have:

``` scala
def extension: Int = foo match {
  case idx if idx != -1 => 15
  case _ => 17
}
```

but compile with `Yinline:false` and it would fail as well. The underlying reason is that, when inlining is enabled, it [performs](https://github.com/scala/scala/blob/5dfcf5eb6a2fea69327420a3ebc4792f9365bf5a/src/compiler/scala/tools/nsc/backend/opt/Inliners.scala#L572) method [normalization](https://github.com/scala/scala/blob/5dfcf5eb6a2fea69327420a3ebc4792f9365bf5a/src/compiler/scala/tools/nsc/backend/icode/Members.scala#L234) that gets rid of trivial `SWITCH` altogether. The constant inlining that runs after inlining doesn't see problematic SWITCH node anymore.

The other way of disabling inlining and method normalization is by marking something `lazy` so the synthetic method generated for `lazy val` looks like this:

``` scala
private def extension$lzycompute(): Int = {
      {
        Si8306.this.synchronized({
          if (Si8306.this.bitmap$0.unary_!())
            {
              Si8306.this.extension = {
                case <synthetic> val x1: Int = Si8306.this.foo();
                (x1: Int) match {
                  case _ => if (x1.!=(-1))
                    15
                  else
                    17
                }
              };
              Si8306.this.bitmap$0 = true;
              ()
            };
          scala.runtime.BoxedUnit.UNIT
        });
        ()
      };
      Si8306.this.extension
    };
```

It seems like the call to `synchronized` prevents inliner from normalizing the method hence `SWITCH` would survive.
